pkg/http/api_directmessaging_test.go

```diff
@@ -0,0 +1,1078 @@
+		apiPath := "v1.0/invoke/http://api.github.com/method/fakeMethod"
+		fakeData := []byte("fakeData")
+		defer fakeInternalErrorResponse.Close()
+
+		mockDirectMessaging.Calls = nil // reset call count
+
+		mockDirectMessaging.
+			On(
+				"Invoke",
+				mock.MatchedBy(matchContextInterface),
+				"http://api.github.com",
+				mock.AnythingOfType("*v1.InvokeMethodRequest"),
+			).
+			Return(fakeInternalErrorResponse, nil).
+			Once()
+
+		// act
+		resp := fakeServer.DoRequest("POST", apiPath, fakeData, nil)
+
+		// assert
+		mockDirectMessaging.AssertNumberOfCalls(t, "Invoke", 1)
+		assert.Equal(t, 500, resp.StatusCode)
+		assert.True(t, strings.HasPrefix(string(resp.RawBody), "{\"errorCode\":\"ERR_MALFORMED_RESPONSE\",\"message\":\""))
+	})
+
+	t.Run("Invoke direct messaging with querystring - 200 OK", func(t *testing.T) {
+		fakeDirectMessageResponse := getFakeDirectMessageResponse()
+		defer fakeDirectMessageResponse.Close()
+
+		apiPath := "v1.0/invoke/fakeAppID/method/fakeMethod?param1=val1&param2=val2"
+		fakeData := []byte("fakeData")
+
+		mockDirectMessaging.Calls = nil // reset call count
+
+		mockDirectMessaging.
+			On(
+				"Invoke",
+				mock.MatchedBy(matchContextInterface),
+				mock.MatchedBy(func(b string) bool {
+					return b == "fakeAppID"
+				}),
+				mock.MatchedBy(func(imr *invokev1.InvokeMethodRequest) bool {
+					if imr == nil {
+						return false
+					}
+
+					msg := imr.Message()
+					if msg.Method != "fakeMethod" {
+						return false
+					}
+
+					if imr.EncodeHTTPQueryString() != "param1=val1&param2=val2" {
+						return false
+					}
+
+					return true
+				}),
+			).
+			Return(fakeDirectMessageResponse, nil).
+			Once()
+
+		// act
+		resp := fakeServer.DoRequest("POST", apiPath, fakeData, nil)
+
+		// assert
+		mockDirectMessaging.AssertNumberOfCalls(t, "Invoke", 1)
+		assert.Equal(t, 200, resp.StatusCode)
+		assert.Equal(t, "fakeDirectMessageResponse", string(resp.RawBody))
+	})
+
+	t.Run("Invoke direct messaging with querystring for external invocation - 200 OK", func(t *testing.T) {
+		fakeDirectMessageResponse := getFakeDirectMessageResponse()
+		defer fakeDirectMessageResponse.Close()
+
+		apiPath := "v1.0/invoke/http://api.github.com/method/fakeMethod?param1=val1&param2=val2"
+		fakeData := []byte("fakeData")
+
+		mockDirectMessaging.Calls = nil // reset call count
+
+		mockDirectMessaging.
+			On(
+				"Invoke",
+				mock.MatchedBy(matchContextInterface),
+				mock.MatchedBy(func(b string) bool {
+					return b == "http://api.github.com"
+				}),
+				mock.MatchedBy(func(imr *invokev1.InvokeMethodRequest) bool {
+					if imr == nil {
+						return false
+					}
+
+					msg := imr.Message()
+					if msg.Method != "fakeMethod" {
+						return false
+					}
+
+					if imr.EncodeHTTPQueryString() != "param1=val1&param2=val2" {
+						return false
+					}
+
+					return true
+				}),
```

Comment:
This is the test that validates the fix (and it's present in a couple of other places)